go/store/datas/database_common.go

```diff
@@ -547,67 +563,186 @@ func (db *database) doUpdateWorkingSet(ctx context.Context, datasetID string, wo
+func (db *database) CommitWithWorkingSet(
+	ctx context.Context,
+	commitDS, workingSetDS Dataset,
+	val types.Value, workingSetSpec WorkingSetSpec,
+	prevWsHash hash.Hash, opts CommitOptions,
+) (Dataset, Dataset, error) {
+	workingSet, err := NewWorkingSet(ctx, workingSetSpec.Meta, workingSetSpec.WorkingRoot, workingSetSpec.StagedRoot, workingSetSpec.MergeState)
+	if err != nil {
+		return Dataset{}, Dataset{}, err
+	}
+
+	err = db.validateWorkingSet(workingSet)
+	if err != nil {
+		return Dataset{}, Dataset{}, err
+	}
+
+	workingSetRef, err := db.WriteValue(ctx, workingSet) // will be orphaned if the tryCommitChunks() below fails
+	if err != nil {
+		return Dataset{}, Dataset{}, err
+	}
+
+	wsValRef, err := types.ToRefOfValue(workingSetRef, db.Format())
+	if err != nil {
+		return Dataset{}, Dataset{}, err
+	}
+
+	commit, err := buildNewCommit(ctx, commitDS, val, opts)
+	if err != nil {
+		return Dataset{}, Dataset{}, err
+	}
+
+	commitRef, err := db.WriteValue(ctx, commit) // will be orphaned if the tryCommitChunks() below fails
+	if err != nil {
+		return Dataset{}, Dataset{}, err
+	}
+
+	commitValRef, err := types.ToRefOfValue(commitRef, db.Format())
+	if err != nil {
+		return Dataset{}, Dataset{}, err
+	}
+
+	var tryCommitErr error
+	testSetFailed := false
+	for tryCommitErr = ErrOptimisticLockFailed; tryCommitErr == ErrOptimisticLockFailed && !testSetFailed; {
+		tryCommitErr = func() error {
+			currentRootHash, err := db.rt.Root(ctx)
 			if err != nil {
 				return err
 			}
 
-			// Second level of locking: assert that the dataset value we read is unchanged from its expected value.
-			// This is separate than the whole-DB lock in the outer loop, as it only protects the value of this dataset
-			// entry. Other writers can update other values and writes chunks in the database and we will retry
-			// indefinitely. But if we find the expected value in the dataset has changed, we immediately abort and the
-			// caller must retry after reconciliation.
-			if ds.HasHead() {
-				head, ok := ds.MaybeHead()
-				if !ok {
-					panic("no head found")
-				}
+			currentDatasets, err := db.DatasetsInRoot(ctx, currentRootHash)
+			if err != nil {
+				return err
+			}
+
+			success, err := db.assertDatasetHash(ctx, currentDatasets, workingSetDS.ID(), prevWsHash)
+			if err != nil {
+				return err
+			}
+
+			if !success {
+				testSetFailed = true
+				return nil
+			}
+
+			r, hasHead, err := currentDatasets.MaybeGet(ctx, types.String(commitDS.ID()))
+			if err != nil {
+				return err
+			}
 
-				h, err := head.Hash(db.Format())
+			// First commit in dataset is always fast-forward, so go through all this iff there's already a Head for datasetID.
+			if hasHead {
+				head, err := r.(types.Ref).TargetValue(ctx, db)
 				if err != nil {
 					return err
 				}
 
-				if h != currHash {
-					testSetFailed = true
-					return ErrOptimisticLockFailed
+				currentHeadRef, err := types.NewRef(head, db.Format())
```

Comment:
This is weird to me...after

```go
v, err := ref.TargetValue(ctx, db)
if err != nil {
    return err
}
r, err := types.NewRef(v, db.Format())
if err != nil {
    return err
}
```

doesn't `r == ref`. Why did I do TargetValue and then re-ref the value?

Reply:
Sometimes you just gotta round trip to the DB for literally no reason